turbopack/crates/turbo-tasks/src/task/function.rs

```diff
@@ -172,22 +166,47 @@ macro_rules! task_inputs_impl {
     }
 }
 
-fn get_args<T: MagicAny>(arg: &dyn MagicAny) -> Result<&T> {
-    let value = arg.downcast_ref::<T>();
-    #[cfg(debug_assertions)]
-    return anyhow::Context::with_context(value, || {
-        format!(
-            "Invalid argument type, expected {} got {}",
-            std::any::type_name::<T>(),
-            (*arg).magic_type_name()
-        )
-    });
-    #[cfg(not(debug_assertions))]
-    return anyhow::Context::context(value, "Invalid argument type");
+// Helper function for `task_fn_impl!()` to reduce the amount of code inside the macro, and to give
+// the compiler more chances to dedupe monomorphized code across small functions with less typevars
+#[cfg(debug_assertions)]
+fn get_debug_downcast_error_msg<T: MagicAny>(arg: &dyn MagicAny) -> String {
+    format!(
+        "Invalid argument type, expected {} got {}",
+        std::any::type_name::<T>(),
+        (*arg).magic_type_name()
+    )
+}
+
+// Helper function for `task_fn_impl!()`
+async fn output_try_into_non_local_raw_vc(output: impl TaskOutput) -> Result<RawVc> {
+    // TODO: Potential future optimization: If we know we're inside a local task, we can avoid
+    // calling `to_non_local()` here, which might let us avoid constructing a non-local cell for the
+    // local task's return value. Flattening chains of `RawVc::LocalOutput` may still be useful to
+    // reduce traversal later.
+    output.try_into_raw_vc()?.to_non_local().await
 }
 
 macro_rules! task_fn_impl {
-    ( $async_fn_trait:ident $arg_len:literal $( $arg:ident )* ) => {
+    ( $async_fn_trait:ident $get_args_fn:ident $arg_len:literal $( $arg:ident )* ) => {
+        /// Downcast, unpack, and clone all the arguments in the singular `arg` tuple.
+        //
+        // This helper is generic over the arguments, but not the function, potentially allowing two
+        // functions with the same argument types to potentially share monomorphized code.
+        #[allow(non_snake_case)]
+        fn $get_args_fn<$($arg: TaskInput + 'static,)*>(arg: &dyn MagicAny) -> Result<($($arg,)*)> {
+            match arg.downcast_ref::<($($arg,)*)>() {
+                Some(($($arg,)*)) => Ok(($($arg.clone(),)*)),
+                None => {
+                    #[cfg(debug_assertions)]
+                    let context_str = get_debug_downcast_error_msg::<($($arg,)*)>(arg);
+                    #[cfg(not(debug_assertions))]
+                    let context_str = "Invalid argument type";
+
+                    anyhow::bail!(context_str);
+                }
+            }
+        }
```

Comment:
Moved this inside the macro, which lets us deduplicate the `.clone()` calls, instead of putting them in each individual trait implementation.